File: frontend/lib/queue-eta.js
// Pure render/format helpers for the worker-queue ETA card.
// All times are approximate; missing data degrades to counts-only text.

export function etaMinutes(seconds) {
  if (seconds == null || Number.isNaN(Number(seconds)) || seconds <= 0) return null;
  return Math.max(1, Math.ceil(Number(seconds) / 60));
}

export function finishClock(finishS, nowMs = Date.now()) {
  if (finishS == null || Number.isNaN(Number(finishS))) return null;
  const d = new Date(nowMs + Number(finishS) * 1000);
  const hh = String(d.getHours()).padStart(2, "0");
  const mm = String(d.getMinutes()).padStart(2, "0");
  return `${hh}:${mm}`;
}

export function renderQueueEta(payload, lang = "ko") {
  if (!payload || typeof payload !== "object") return "";
  const ko = String(lang || "").startsWith("ko");
  const perStage = Array.isArray(payload.per_stage) ? payload.per_stage : [];
  const queued = perStage.reduce((a, s) => a + (Number(s.queued) || 0), 0);
  const running = perStage.reduce((a, s) => a + (Number(s.running) || 0), 0);

  if (payload.fallback || payload.est_finish_s == null) {
    return ko
      ? `대기 ${queued} · 실행 ${running} (예상 시간 산출 불가)`
      : `queued ${queued} · running ${running} (estimate unavailable)`;
  }

  const mins = etaMinutes(payload.est_finish_s);
  const clock = finishClock(payload.est_finish_s);
  const stage = payload.current_stage || (perStage[0] && perStage[0].stage) || "-";
  return ko
    ? `현재 단계 ${stage} · 내 앞 대기 ${queued} · 예상 완료 ~${mins}분 후(~${clock}, 근사)`
    : `stage ${stage} · ${queued} ahead · ~${mins} min (~${clock}, approx)`;
}

File: pipeline-mcp/src/pipeline_mcp/queue_eta.py
"""Stateless queue-ETA math.

Pure functions turning (jobs_ahead, workers, avg_duration) into approximate
wait/finish seconds, and summing per-stage estimates into a whole-run estimate.
No I/O, no RunPod. All estimates are explicitly approximate; missing duration
data degrades to a fallback (counts-only) with no time.
"""
from __future__ import annotations

import math
from typing import Any


def estimate_stage_eta(
    *, jobs_ahead: int, workers: int, avg_duration_s: float | None
) -> dict[str, Any]:
    if avg_duration_s is None or avg_duration_s <= 0:
        return {"wait_s": None, "finish_s": None, "approximate": True, "fallback": True}
    ahead = max(int(jobs_ahead), 0)
    w = max(int(workers), 1)
    wait_s = math.ceil(ahead / w) * float(avg_duration_s)
    return {
        "wait_s": wait_s,
        "finish_s": wait_s + float(avg_duration_s),
        "approximate": True,
        "fallback": False,
    }


def estimate_run_eta(stages: list[dict[str, Any]]) -> dict[str, Any]:
    known = [s for s in stages if not s.get("fallback") and s.get("finish_s") is not None]
    any_fallback = any(s.get("fallback") for s in stages)
    est_finish = sum(float(s["finish_s"]) for s in known) if known else None
    return {
        "est_finish_s": est_finish,
        "approximate": True,
        "fallback": bool(any_fallback),
        "per_stage": stages,
    }
